File: solutions/173. Binary Search Tree Iterator/173.py
class BSTIterator:
  def __init__(self, root: TreeNode | None):
    self.i = 0
    self.vals = []
    self._inorder(root)

  def next(self) -> int:
    self.i += 1
    return self.vals[self.i - 1]

  def hasNext(self) -> bool:
    return self.i < len(self.vals)

  def _inorder(self, root: TreeNode | None) -> None:
    if not root:
      return
    self._inorder(root.left)
    self.vals.append(root.val)
    self._inorder(root.right)

File: solutions/173. Binary Search Tree Iterator/173.java
class BSTIterator {
  public BSTIterator(TreeNode root) {
    inorder(root);
  }

  public int next() {
    return vals.get(i++);
  }

  public boolean hasNext() {
    return i < vals.size();
  }

  private int i = 0;
  private List<Integer> vals = new ArrayList<>();

  private void inorder(TreeNode root) {
    if (root == null)
      return;
    inorder(root.left);
    vals.add(root.val);
    inorder(root.right);
  }
}
